### PR TITLE
changed structure + pytests

### PR DESCRIPTION
**Pull request by arnavdas88**
Signed-off-by: Arnav Das <arnav.das88@gmail.com>

changed structure  + unit tests

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

Added Gitignore

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

install_requires -> versions specified

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

Update LICENSE

noxfile added

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

setup numpy version

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

license-header.py

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

Test general_functions

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

structuring general functions and states

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

general_functions removed files

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

License Added to all files

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

states removed files

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

np.matrix deprication

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

tests for general_functions and states

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

states imported norm() from np.linalg

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

nox tests -> test

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

structuring linalg

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

nox update

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

setup.py update

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

gitignore update

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

structuring distance_measures

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

tests for distance_measures and linalg

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

file header cleanup

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

structuring gates

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

structured pauli

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

tests for gates and pauli

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

structuring Weyl

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

np.int deprication

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

restructuring fermions

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

restructuring su

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

tests for su, weyl, fermions

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

some comments

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

restructuring clifford and fidility

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

tests clifford and fidilities

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

setup.py updated

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

restructuring entropies

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

restructuring channels

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

restructuring protocols

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

tests for entropies

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

[WIP] tests for channels & protocols

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

noxfile update + timing function added

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

All Tests

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

states.utils

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

All tests completed

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

update

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

Create ci.yml

Update ci.yml

requirements

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

pauli fix

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

update ci.yml

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

entropies test update

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

test + nox

Signed-off-by: Arnav Das <arnav.das88@gmail.com>

### CODE REVIEW

File: qutipy/Clifford/__init__.py [hunk 1]
```diff
@@ -1,4 +1,250 @@
...
+
+
+def Clifford_twirl_channel_one_qubit(K, rho, sys=1, dim=[2]):
+    """
+    Twirls the given channel with Kraus operators in K by the one-qubit
+    Clifford group on the given subsystem (specified by sys).
+    """
+
+    n = int(np.log2(np.sum([d for d in dim])))
+
```
Comment (sumeetkhatri) [inline qutipy/Clifford/__init__.py:65]: All multiplications in this function need to changed from '*' -> '@'. Make sure Rx_i, Ry_i, Rz_i are returning np.array.

File: qutipy/Pauli/__init__.py [hunk 2]
```diff
@@ -1,8 +1,220 @@
...
+            C.append(Tr(dag(sigma_i) @ X))
+
+    return C
+
+
+def nQubit_quadratures(n):
+    """
+    Returns the list of n-qubit "quadrature" operators, which are defined as
+    (for two qubits)
+
```
Comment (sumeetkhatri) [inline qutipy/Pauli/__init__.py:169]: Replace all '\otimes' with the tensor product symbol.

File: qutipy/general_functions/__init__.py [hunk 3]
```diff
@@ -1,16 +1,487 @@
...
+
+    ket(d,[j1,j2,...,jn]) generates the tensor product |j1>|j2>...|jn> of
+    d-dimensional basis vectors.
+
+    If dim is specified as a list, then, e.g., ket([d1,d2],[j1,j2]) generates the
+    tensor product |j1>|j2>, with the first tensor factor being d1-dimensional
+    and the second tensor factor being d2-dimensional.
+    """
+
+    # TODO: Dimensions are probably not required
```
Comment (sumeetkhatri) [inline qutipy/general_functions/__init__.py:143]: In general, the dimensions are required, because it affects the size of the array at the output.

**Review by sumeetkhatri — COMMENTED** [review]
Some comments have been given.

File: qutipy/Clifford/__init__.py (hunk 1 above)
Comment (arnavdas88) [reply qutipy/Clifford/__init__.py]: Refer to https://github.com/arnavdas88/QuTIpy/blob/master/qutipy/clifford.py#L58

File: qutipy/general_functions/__init__.py (hunk 3 above)
Comment (arnavdas88) [reply qutipy/general_functions/__init__.py]: Fixed. Refer to https://github.com/arnavdas88/QuTIpy/blob/master/qutipy/general_functions.py#L125

File: qutipy/Pauli/__init__.py (hunk 2 above)
Comment (arnavdas88) [reply qutipy/Pauli/__init__.py]: Fixed. Refer to https://github.com/arnavdas88/QuTIpy/blob/master/qutipy/pauli.py#L165

**Comment by arnavdas88** [comment]
I have committed some updates... Please go through them once. 😁

**Review by sumeetkhatri — APPROVED** [review]

**Pull request merged**